### PR TITLE
Change the way we calculate uuids.

### PR DESCRIPTION
**Pull request by ralphiee22**
## Summary

We now calculate uuids based on the partition values of the model, as well as the source id (which will be unique fields on the model or a random uuid). 

NOTE:
Tests will break, but we will fix at a later time.

### CODE REVIEW

File: morango/models.py [hunk 1]
```diff
@@ -238,13 +245,19 @@ class SyncableModel(UUIDModelMixin):
     their data syncable across devices.
     """
 
+    # constant value to insert into partition strings in place of current model's ID, as needed (to avoid circularity)
+    ID_PLACEHOLDER = "<id>"
```
Comment (jamalex) [inline morango/models.py]: Let's actually replace this with `${id}`, so it follows the same `Template` format as `*_scope_def` strings.

File: morango/models.py [hunk 2]
```diff
@@ -238,13 +245,19 @@ class SyncableModel(UUIDModelMixin):
...
     _morango_internal_fields_not_to_serialize = ('_morango_dirty_bit',)
     morango_fields_not_to_serialize = ()
+    morango_profile = None
 
     # morango specific field used for tracking model changes
     _morango_dirty_bit = models.BooleanField(default=True)
     # morango specific field used to store random uuid or unique together fields
     _morango_source_id = models.CharField(max_length=96)
+    # morango specific field used to store the partition on the model
+    _morango_partition = models.TextField()
```
Comment (jamalex) [inline morango/models.py]: `CharField` should be enough for this -- max_length 128, perhaps?

File: morango/models.py [hunk 3]
```diff
@@ -290,18 +303,23 @@ def deserialize(cls, dict_model):
     def merge_conflict(cls, current, incoming):
         return incoming
 
-    def get_partition(self, *args, **kwargs):
-        """Should return a dictionary with any relevant partition keys included, along with their values."""
-        raise NotImplemented("You must define a 'get_partition_names' method on models that inherit from SyncableModel.")
-
+    def calculate_source_id(self):
+        """Should return a string that uniquely defines the model or `None` for a random uuid."""
```
Comment (jamalex) [inline morango/models.py]: "model" -> "model instance"

File: tests/testapp/facility_profile/models.py [hunk 4]
```diff
@@ -32,9 +32,11 @@ class Facility(MorangoMPTTModel, FacilityDataSyncableModel):
     now_date = models.DateTimeField(default=timezone.now)
     parent = TreeForeignKey('self', null=True, blank=True, related_name='children', db_index=True)
 
-    def get_partition(self, *args, **kwargs):
-        return {}
+    def calculate_source_id(self, *args, **kwargs):
+        return ''
```
Comment (jamalex) [inline tests/testapp/facility_profile/models.py]: Why empty string?

Comment (jamalex) [inline tests/testapp/facility_profile/models.py]: (or is that just temporary until the tests are fixed?)

Comment (ralphiee22) [inline tests/testapp/facility_profile/models.py]: temp until tests are fixed.

**Pull request merged**